### ISSUE TITLE
[python-package] When using custom metrics, num_threads setting is not working in lightgbm?

### ISSUE THREAD

**Issue by JisongXie** [microsoft/LightGBM#5624]
## Description

When I train a lightgbm model with my own custom metrics, I find that model is trained in single thread, though I set `"num_threads": 16`.

    params = {
      ...,
      "num_threads": 16,
      ...
    }
    lgb.train(params,
        feval=[my_own_metrics],
        ...
    )

However,  when I use the metrics (e.g. auc) provided by lightgbm, it could run in multi-threading as expected.

    params = {
      ...,
      "num_threads": 16,
      "metrics": ["auc"],
      ...
    }

## Reproducible example
I use lightgbm in jupyter and dokcer.

My own metric implementation is as follow:

```
def my_own_metrics(y_pred_arr, data):
    y_true_arr = data.get_label()
    
    y_true = pd.DataFrame(y_true_arr, columns=['target'])
    y_pred = pd.DataFrame(y_pred_arr, columns=['prediction'])
    
    def top_four_percent_captured(y_true: pd.DataFrame, y_pred: pd.DataFrame) -> float:
        df = (pd.concat([y_true, y_pred], axis='columns')
              .sort_values('prediction', ascending=False))
        df['weight'] = df['target'].apply(lambda x: 20 if x==0 else 1)
        four_pct_cutoff = int(0.04 * df['weight'].sum())
        df['weight_cumsum'] = df['weight'].cumsum()
        df_cutoff = df.loc[df['weight_cumsum'] <= four_pct_cutoff]
        return (df_cutoff['target'] == 1).sum() / (df['target'] == 1).sum()
        
    def weighted_gini(y_true: pd.DataFrame, y_pred: pd.DataFrame) -> float:
        df = (pd.concat([y_true, y_pred], axis='columns')
              .sort_values('prediction', ascending=False))
        df['weight'] = df['target'].apply(lambda x: 20 if x==0 else 1)
        df['random'] = (df['weight'] / df['weight'].sum()).cumsum()
        total_pos = (df['target'] * df['weight']).sum()
        df['cum_pos_found'] = (df['target'] * df['weight']).cumsum()
        df['lorentz'] = df['cum_pos_found'] / total_pos
        df['gini'] = (df['lorentz'] - df['random']) * df['weight']
        return df['gini'].sum()

    def normalized_weighted_gini(y_true: pd.DataFrame, y_pred: pd.DataFrame) -> float:
        y_true_pred = y_true.rename(columns={'target': 'prediction'})
        return weighted_gini(y_true, y_pred) / weighted_gini(y_true, y_true_pred)

    g = normalized_weighted_gini(y_true, y_pred)
    d = top_four_percent_captured(y_true, y_pred)

    return  'metric', 0.5 * (g + d), True
```

## Environment info

LightGBM version or commit hash:  3.3.2

Command(s) you used to install LightGBM

```
pip install lightgbm
```

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

Can you please describe specifically what evidence you have the "num_threads is not working"?

**Comment by jameslamb** [comment]
I agree completely with @jmoralez 's assessment.

>  It seems than the num_threads parameter only work in training phase, while not in evaling phase.

This is not accurate.

The `num_threads` parameter in LightGBM control's parallelism in **lightgbm's** own code. When you provide your own custom Python code for an evaluation metric, LightGBM won't parallelize execution of that function. It just passes **all of the labels and predictions**, and it's you're responsibility to control what that function does and how it executes.

That's by design. LightGBM can't assume that all custom metric functions could be parallelized (e.g. evaluated concurrently over non-overlapping sections of the data). Some people use metric functions that require global information about the training data, e.g. the median of the predicted values in regression.

**Comment by JisongXie** [comment]
Thanks for your explanation @jameslamb, I totally agree with you.
